### ISSUE TITLE
Nested conditional types with type parameters extending union are not inferred

### ISSUE THREAD

**Issue by cleydyr** [microsoft/TypeScript#59127]
### 🔎 Search Terms

* nested conditional types
* parameterized types
* union types

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about conditional types.

### ⏯ Playground Link

https://www.typescriptlang.org/play/?#code/C4TwDgpgBAggQlAvFA5AQxVAPqgRpnFAYxQChTRIoBpAHgBUoIAPYCAOwBMBnWOAPiRRGLNl17pMAfijsArgFtcEAE5QAXMKasOPPNNmLlazfKWryAMznsiwAJYB7dlEuPHtAKraxe+PwAKAEpNOk9BAG9SKCgVCGA5FRcAFgAmUgBfIA

### 💻 Code

```ts
type AB = 'a' | 'b' | 'c'

type K<T extends AB> = T extends 'a' ? number : T extends 'b' ? number : number

function foo<U extends AB>(): K<U> {
  return 42
}
```

### 🙁 Actual behavior

TS gives an error: ```Type 'number' is not assignable to type 'K<U>'.```

### 🙂 Expected behavior

TS should not throw an error and guess  as it doesn't give when the conditional is simpler:

### Additional information about the issue

I thought it had to do with [distributive conditionals](https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#distributive-conditional-types). Wrapping the type parameters in square brackets makes the error disappear but I don't understand why.

**Comment by ConnorDuvalS** [comment]
At a guess, this is specifically to handle the case that `any` is used in the generic (as `any extends _ ? A : B` would just be `A | B`). I am unsure why, for discrete unions, such as this, the type is not wholly evaluated. This is likely an efficiency concern but it certainly does not behave as I would expect.

What's weird is that `K<T>` will always extend `number`.


```ts
type Extension<T, U extends T> = U;

type ABC = "a" | "b" | "c";
type K<T extends ABC> = T extends "a" ? number : T extends "b" ? number : number;
type L<T extends ABC> = Extension<number, K<T>>;

type A = L<"a">;
type B = L<"a">;
type C = L<"a">;
```

I am hoping this sees some interest from someone familiar with the TypeScript codebase.

**Comment by RyanCavanaugh** [comment]
We don't check to see if a type would be assignable to every leaf node of an unresolved conditional type, because if it was, you wouldn't have needed to write a conditional type in the first place. In other words there's no real difference from a caller's perspective between `foo` as written and

```ts
function foo<U extends AB>(): number {
  return 42
}
```

**Comment by cleydyr** [comment]
> We don't check to see if a type would be assignable to every leaf node

Unless the tree is as simple as in this case?

```typescript
type AB = 'a' | 'b'

type K<T extends AB> = T extends 'a' ? number : number

function foo<U extends AB>(): K<U> {
  return 42
}
```

In the case above, Typescript infers that the type can be `number` or `number`, so `number` only, and no TS error is shown. Is that an exception to your statement or did I misunderstand it?
